### PR TITLE
[NYUKI-221] Global cleanup of nyuki's code and librairies

### PR DESCRIPTION
**Pull request by gdraynz**

### CODE REVIEW

File: nyuki/logs.py [hunk 1]
```diff
@@ -1,41 +1,29 @@
...
-        "websockets": {
-            "level": "WARNING"
-        }
     },
-    "disable_existing_loggers": False
+    'root': {'handlers': ['console'], 'level': 'INFO'},
+    'loggers': {
+        'asyncio': {'handlers': ['loop_stderr'], 'level': 'WARNING'},
+        'gmqtt': {'level': 'WARNING'},
+        'slixmpp': {'level': 'WARNING'},
```
Comment (lhavel) [inline nyuki/logs.py]: As far as I know we no longer use `slixmpp`

File: nyuki/config.py [hunk 2]
```diff
@@ -47,66 +24,77 @@ def merge_configs(*configs):
...
+                'File %s does not exist, using default.json',
+                self._location,
+            )
+            self._location = self.DEFAULT_CONF_FILE
+            conf = json.loads(self._location.read_text())
+        return merge_configs(self.BASE_CONF, conf)
+
+    def _get_from_s3(self):
+        """
+        TODO: Use boto3 ?
```
Comment (lhavel) [inline nyuki/config.py]: Indeed, use boto3 instead of the awscli. I found a snippet in seconds:
```python
import boto3

s3 = boto3.resource('s3')

obj = s3.Object(bucket, key)
obj.get()['Body'].read().decode('utf-8')
```

File: nyuki/nyuki.py [hunk 3]
```diff
@@ -35,62 +32,30 @@ class Nyuki:
...
-        bus_config = self._config.get('bus')
-        if bus_config is not None:
-            bus_service = bus_config.get('service', 'mqtt')
-            if bus_service == 'mqtt':
-                self._services.add('bus', MqttBus(self))
-
-        self.is_stopping = False
+        self.services = ServiceManager(self)
+        self.services.add('api', Api(self))
+        self.services.add('bus', MqttBus(self))
```
Comment (lhavel) [inline nyuki/nyuki.py]: It's convenient to start a nyuki standalone for testing/dev without even a broker.
If there's no `bus` key in the config, the nyuki should not event try to connect.

File: nyuki/nyuki.py [hunk 4]
```diff
@@ -100,196 +65,126 @@ def __getattribute__(self, name):
...
+        await nyuki.start()
+        await nyuki.wait()
 
+    async def start(self):
+        """
+        Start the nyuki's services.
+        """
+        self.__stop_event.clear()
+        # Set signal handlers
+        loop = asyncio.get_event_loop()
```
Comment (lhavel) [inline nyuki/nyuki.py]: use the `self.loop` property here

File: nyuki/nyuki.py [hunk 5]
```diff
@@ -35,62 +32,30 @@ class Nyuki:
...
+        self._configuration = Configuration(
+            command_args.config or os.environ.get('CONF_LOCATION'),
+        )
+
+        self._log_level = command_args.logging
+        self.configure_logging(DEFAULT_LOGGING)
+        self.configure()
 
         # Setup stack sampling
         self._sampler = None
```
Comment (lhavel) [inline nyuki/nyuki.py]: It looks like this attribute `self._sampler` is not used anywhere in the code...

File: nyuki/nyuki.py [hunk 6]
```diff
@@ -100,196 +65,126 @@ def __getattribute__(self, name):
...
+        # Set logging level
+        if self._log_level:
+            logging.getLogger().setLevel(self._log_level)
 
     async def setup(self):
         """
         First thing called when starting the event loop, coroutine or not.
         """
-        log.warning('Setup called, but not overridden')
+        log.debug('Setup called, but not overridden')
```
Comment (lhavel) [inline nyuki/nyuki.py]: The warning level was fine

File: nyuki/utils/__init__.py [hunk 7]
```diff
@@ -1,4 +1,9 @@
+import os
+
 from .dtutils import from_isoformat, utcnow
 from .evaluate import safe_eval, ConditionBlock
-from .serialize import serialize_object
+from .serialize import serialize_object, json_default
 from .transform import Converter
+
+
+IS_DEBUG = os.environ.get('DEBUG') == '1'
```
Comment (lhavel) [inline nyuki/utils/__init__.py]: The log level should be all that must be defined

File: nyuki/services.py [hunk 8]
```diff
@@ -5,24 +5,23 @@
 log = logging.getLogger(__name__)
 
 
-class Service(object):
+class Service:
 
     async def start(self, *args, **kwargs):
-        raise NotImplementedError
+        pass
```
Comment (lhavel) [inline nyuki/services.py]: Why did you remove the `raise NotImplementedError`; are there situations where one of these methods might not be overridden?

File: nyuki/api/api.py [hunk 9]
```diff
@@ -16,201 +14,78 @@
...
-        except HTTPBreak as exc:
-            return Response(exc.body, status=exc.status)
-
-        if isinstance(capa_resp, Response):
-            return capa_resp
-        return Response()
-
-    return middleware
-
+            return await handler(**self.request.match_info)
```
[the remark this answers is not in the document]
Comment (gdraynz) [reply nyuki/api/api.py]: This allows to create and expand a custom api error handler using the `singledispatch` method, and return JSON errors more easily.

File: nyuki/bus/mqtt.py [hunk 10]
```diff
@@ -62,64 +37,41 @@ def topics(self):
...
-            except asyncio.CancelledError:
-                log.debug('future cancelled: %s', future)
-        self.connect_future = asyncio.ensure_future(self._run())
-        self.connect_future.add_done_callback(cancelled)
+                await self.client.connect(
+                    self._dsn.host,
+                    self._dsn.port,
+                    (self._dsn.scheme == 'mqtts'),
+                )
+            except ConnectionRefusedError as exc:
```
[the remark this answers is not in the document]
Comment (gdraynz) [reply nyuki/bus/mqtt.py]: I remember talks about nyukis that should never stop if a third-party service is down but simply try to reconnect but yes stopping it after a while may be better

File: nyuki/config.py (hunk 2 above)
Comment (gdraynz) [reply nyuki/config.py]: The thing is this method is not really intended to be used right now as we're kind of between two ways of configuring our nyukis and I didn't want to add the (big) boto3 dependency right now.

File: nyuki/nyuki.py (hunk 6 above)
Comment (gdraynz) [reply nyuki/nyuki.py]: This is purely an implementation thing that should be relevant only when developing a nyuki and not during its normal execution.

File: nyuki/services.py (hunk 8 above)
Comment (gdraynz) [reply nyuki/services.py]: Same as above, this is an implementation specificity and some services (like the StackSampler here) may not need to implement these 3 methods, yet it's easier to call the 3 methods on all available services at once. I added a debug log to simply notice this.

File: nyuki/nyuki.py (hunk 6 above)
Comment (lhavel) [reply nyuki/nyuki.py]: Then it should be fixed during the development phase and never happen in production.
The debug level just hides real problems or things we might have missed during the dev.

File: nyuki/utils/__init__.py (hunk 7 above)
Comment (lhavel) [reply nyuki/utils/__init__.py]: As far as I can see it is used only to insert exception details into the JSON response in "DEBUG mode". Which is not a good practice. The logs in DEBUG level are best suited for that and does not cost more.

File: nyuki/api/errors.py [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+from nyuki.utils import IS_DEBUG
+
+
+log = logging.getLogger(__name__)
+
+
+@singledispatch
+def api_error_handler(exc, request):
+    log.exception(exc)
+    if IS_DEBUG:
```
[the remark this answers is not in the document]
Comment (gdraynz) [reply nyuki/api/errors.py]: Exceptions are always logged, but will also be showed in the HTTP response if the nyuki runs in debug

File: nyuki/nyuki.py (hunk 6 above)
Comment (gdraynz) [reply nyuki/nyuki.py]: No these are methods that are always called by the nyuki library, but not necessarily used by every nyuki. If my nyuki does not have any initial setup or teardown cleaunp to do, I don't care about this warning message popping every time.

File: nyuki/utils/__init__.py (hunk 7 above)
Comment (gdraynz) [reply nyuki/utils/__init__.py]: This is also a small helper to add all sorts of things when developing a nyuki that have some debug specifics, instead of having to fetch the environment variable by hand.

File: nyuki/api/errors.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (lhavel) [reply nyuki/api/errors.py]: Please remove exception details from HTTP response this is definitely a very bad practice.
APIs should never expose implementation details.

File: nyuki/config.py (hunk 2 above)
Comment (lhavel) [reply nyuki/config.py]: Ok then remove this method: why should we keep code that is not used. It just make the codebase harder to maintain.

File: nyuki/nyuki.py (hunk 6 above)
Comment (lhavel) [reply nyuki/nyuki.py]: ok let's keep it like that
